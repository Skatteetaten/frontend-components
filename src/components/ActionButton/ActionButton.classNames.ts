import { mergeStyles } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { FontSizes } from '..';
import { getFocusStyle } from '..';
import { PaletteProps } from '..';
// @ts-ignore TODO
function getTypeColor(props): string {
  const palette = getTheme().palette as PaletteProps;
  switch (props.color) {
    case 'blue':
      // @ts-ignore TODO
      return {
        color: palette.skeColor.blue
      };
    case 'white':
      // @ts-ignore TODO
      return {
        color: palette.skeColor.white
      };
    default:
      // @ts-ignore TODO
      return {
        color: palette.bodyText
      };
  }
}
// @ts-ignore TODO
function getIconColor(props): string {
  const palette = getTheme().palette as PaletteProps;
  switch (props.color) {
    case 'black':
      // @ts-ignore TODO
      return {
        color: palette.bodyText
      };
    case 'green':
      // @ts-ignore TODO
      return {
        color: palette.skeColor.green
      };
    case 'red':
      // @ts-ignore TODO
      return {
        color: palette.skeColor.pink
      };
    case 'white':
      // @ts-ignore TODO
      return {
        color: palette.skeColor.white
      };
    default:
      // @ts-ignore TODO
      return {
        color: palette.skeColor.blue
      };
  }
}
// @ts-ignore TODO
function getHoverColor(props): object {
  const palette = getTheme().palette as PaletteProps;
  switch (props.color) {
    case 'black':
      return {
        color: palette.skeColor.blue
      };
    case 'white':
      return {
        color: palette.skeColor.whiteGrey
      };
    default:
      return {
        color: palette.bodyText
      };
  }
}
// @ts-ignore TODO
export function getClassNames(props): string {
  const palette = getTheme().palette as PaletteProps;
  const inset = -4;
  const radius = 0;

  return mergeStyles([
    // @ts-ignore TODO
    getFocusStyle({ palette }, inset, 'relative', radius),
    {
      selectors: {
        '&.ms-Button.ms-Button--action, &.ms-Button--action .ms-Button-icon': {
          // @ts-ignore TODO
          ...getTypeColor(props)
        },
        '&.ms-Button--action .ms-Button-icon': {
          // @ts-ignore TODO
          fontSize: FontSizes[props.iconSize],
          // @ts-ignore TODO
          ...getIconColor(props)
        },
        '&.ms-Button--action:hover .ms-Button-label': {
          textDecoration: 'underline',
          ...getHoverColor(props)
        },
        '&.ms-Button--action:hover .ms-Button-icon': {
          ...getHoverColor(props)
        },
        '&.ms-Button--action:active .ms-Button-label': {
          textDecoration: 'none'
        },
        '&.ms-Button--action:disabled, &.ms-Button--action:disabled i': {
          color: palette.skeColor.lightGrey
        }
      }
    }
  ]);
}
