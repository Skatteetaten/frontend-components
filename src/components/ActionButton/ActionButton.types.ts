import { ButtonProps } from '../Button';
import { ReactNode } from 'react';

export interface ActionButtonProps extends ButtonProps {
  /** Ikon som skal vises foran lenketeksten */
  icon?: string;
  /** Ikon størrelse, to tilgjengelige størrelser */
  iconSize?: any;
  /** Fire forhåndsdefinerte farger, se eksempler */
  color?: 'blue' | 'black' | 'red' | 'green' | 'white';
  /**  true hvis ikonet skal plasseres etter tekst, ellers rendres det foran. */
  iconAfter?: boolean;
  /** Om knappen skal ha strek under */
  border?: boolean;
  ariaLabel?: string;
  children?: ReactNode;
}
