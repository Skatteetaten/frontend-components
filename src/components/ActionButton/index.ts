export { ActionButton as default } from './ActionButton';
export * from './ActionButton';
