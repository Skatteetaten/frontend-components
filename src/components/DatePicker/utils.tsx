export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
