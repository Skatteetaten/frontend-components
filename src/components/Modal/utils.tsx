import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
