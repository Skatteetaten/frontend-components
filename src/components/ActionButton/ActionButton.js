import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { ActionButton as FabricActionButton } from 'office-ui-fabric-react/lib-commonjs/Button';
import { getClassNames } from './ActionButton.classNames';

/**
 * @visibleName ActionButton (Aksjonsknapp)
 */
export default class ActionButton extends React.PureComponent {
  static NORMAL = 'icon';
  static LARGE = 'xxLarge';

  static propTypes = {
    /** Tekst som er ønskelig at leses opp for skjermleser */
    ariaLabel: PropTypes.string,
    /** Callback for klikk på knappen */
    onClick: PropTypes.func,
    /** Ikon som skal vises foran lenketeksten */
    icon: PropTypes.string,
    /** Ikon størrelse, to tilgjengelige størrelser */
    iconSize: PropTypes.oneOf([ActionButton.NORMAL, ActionButton.LARGE]),
    /** Fire forhåndsdefinerte farger, se eksempler*/
    color: PropTypes.oneOf(['blue', 'black', 'red', 'green', 'white']),
    className: PropTypes.string,
    /** Viser knappen som inaktiv */
    disabled: PropTypes.bool
  };

  static defaultProps = {
    onClick: undefined,
    icon: undefined,
    iconSize: ActionButton.NORMAL,
    color: 'blue',
    disabled: false
  };

  render() {
    const {
      children,
      onClick,
      icon,
      iconSize,
      color,
      className,
      disabled,
      ...props
    } = this.props;
    return (
      <FabricActionButton
        {...props}
        className={classnames(getClassNames(this.props), className)}
        onClick={onClick}
        role="button"
        color={color}
        iconSize={iconSize}
        iconProps={{
          iconName: icon
        }}
        disabled={disabled}
      >
        {children}
      </FabricActionButton>
    );
  }
}
