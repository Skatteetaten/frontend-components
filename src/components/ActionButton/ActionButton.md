** Aksjonsknapp for selvstendige og mindre fremtredende handlinger **

```js
import ActionButton from '@skatteetaten/frontend-components/ActionButton';

<div>
  <ActionButton ariaLabel={'legg-til'} icon="AddOutline">
    Legg til
  </ActionButton>
  <ActionButton ariaLabel={'oppdater'} icon="Update" color="black">
    Oppdater
  </ActionButton>
</div>;
```

Ikonet kan rendres til høyre for teksten med iconAfter=_true_:

```js
import ActionButton from '@skatteetaten/frontend-components/ActionButton';

<div>
  <ActionButton ariaLabel={'legg-til'} icon="delete" iconAfter={true}>
    Fjern
  </ActionButton>
</div>;
```

```js noeditor uu
<h3>Tips</h3>

<ul>
  <li>Bruk korte, forståelige tekster i knappen. Det skal være lett å skjønne hva målet med knappen er. </li>
  <li>Sjekk at ikonet er plassert ved siden av teksten</li>
</ul>

<h3>Mest relevante WCAG-krav</h3>

<ul>
  <li>4.1.2 A, Navn, rolle, verdi</li>
</ul>

<h3>WAI-ARIA</h3>

<ul>
  <li>Aria-hidden="true" brukes på ikonet for skjule det for skjermleser.</li>
  <li>Aria-label kan benyttes på button-tag hvis det er behov for å tydeliggjøre hva knappen gjør.</li>
</ul>
```

```js noeditor beskrivelse

<h3>En knapp for mindre fremtredende handlinger</h3>
<p>
  Aksjonsknapp bruker du for selvstendige og mindre fremtredende handlinger på siden, som for eksempel «Legg til» eller «Fjern». Knappen tar mindre plass og er mer plasseringsvennlig enn <a href='#button'>Button</a>.
</p>
<h3>Ulike varianter av aksjonsknapper</h3>
<p>
  Du skal normalt bruke blå farge på alle klikkbare elementer, også aksjonsknapper. Hvis det er mange blå elementer på siden, kan brukeren oppleve det støyende. Du kan da velge nedtonede aksjonsknapper i sort i stedet. Dersom aksjonsknappen skal være mer synlig kan du øke ikonstørrelsen.
</p>
<p>
  Du kan gi ikonet farge hvis du vil understreke en mening, for eksempel grønn for godkjenn eller rød for avvis.
</p>
<p>
  Se <a href='https://www.skatteetaten.no/stilogtone/skrive/skriveregler/knapper/'>skrivereglene</a> for hvordan skrive på knapper.
</p>

```
