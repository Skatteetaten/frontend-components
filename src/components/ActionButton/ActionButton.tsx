import classnames from 'classnames';
import {
  ActionButton as FabricActionButton,
  IButtonProps
} from 'office-ui-fabric-react/lib-commonjs/Button';
import * as React from 'react';
import { getClassNames } from './ActionButton.classNames';

interface ActionButtonProps extends IButtonProps {
  /** Ikon som skal vises foran lenketeksten */
  icon?: string;
  /** Ikon størrelse, to tilgjengelige størrelser */
  iconSize?: any;
  /** Fire forhåndsdefinerte farger, se eksempler */
  color?: 'blue' | 'black' | 'red' | 'green' | 'white';
}
/**
 * @visibleName ActionButton (Aksjonsknapp)
 */
export default class ActionButton extends React.PureComponent<
  ActionButtonProps,
  {}
> {
  static NORMAL = 'icon';
  static LARGE = 'xxLarge';

  static defaultProps = {
    color: 'blue',
    disabled: false,
    icon: undefined,
    iconSize: ActionButton.NORMAL,
    onClick: undefined
  };

  render() {
    const { children, icon, iconSize, color, className, ...props } = this.props;
    return (
      <FabricActionButton
        {...props}
        className={classnames(getClassNames(this.props), className)}
        role="button"
        color={color}
        iconProps={{
          iconName: icon
        }}
      >
        {children}
      </FabricActionButton>
    );
  }
}
