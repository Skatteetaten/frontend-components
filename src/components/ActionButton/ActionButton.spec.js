import React from 'react';
import { mount } from 'enzyme';
import { matches } from './../utils/test-utils';
import ActionButton from './ActionButton';

function oppsettMount(props) {
  return mount(<ActionButton {...props} />);
}

describe('ActionButton komponent', () => {
  it('matcher snapshot', () => {
    matches(<ActionButton ariaLabel="Button label" />);
  });

  it('rendrer actionbutton med standard props', () => {
    const wrapper = oppsettMount({
      ariaLabel: 'Button label'
    });

    expect(wrapper.prop('iconSize')).toEqual('icon');
    expect(wrapper.prop('color')).toEqual('blue');
    expect(wrapper.prop('disabled')).toEqual(false);
  });

  it('oppdaterer actionbutton med riktig props', () => {
    const wrapper = oppsettMount({
      iconSize: ActionButton.LARGE,
      color: 'green',
      icon: 'AddOutline',
      disabled: true,
      ariaLabel: 'Button label'
    });

    expect(wrapper.prop('iconSize')).toEqual('xxLarge');
    expect(wrapper.prop('color')).toEqual('green');
    expect(wrapper.prop('icon')).toEqual('AddOutline');
    expect(wrapper.prop('disabled')).toEqual(true);
  });
});
