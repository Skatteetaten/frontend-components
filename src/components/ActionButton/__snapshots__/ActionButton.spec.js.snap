// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton komponent matcher snapshot 1`] = `
<button
  aria-label="Button label"
  className=
      ms-Button
      ms-Button--action
      ms-Button--command
      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        background-color: transparent;
        border-radius: 2px;
        border: 1px solid transparent;
        box-sizing: border-box;
        color: rgba(0, 0, 0, 0.95);
        cursor: pointer;
        display: inline-block;
        font-family: 'Helvetica Neue', Helvetica , Arial, 'Liberation Sans', 'Nimbus Sans L', sans-serif;
        font-size: 1rem;
        font-weight: 400;
        height: 40px;
        outline: transparent;
        padding-bottom: 0;
        padding-left: 4px;
        padding-right: 4px;
        padding-top: 0;
        position: relative;
        text-align: center;
        text-decoration: none;
        user-select: none;
      }
      &::-moz-focus-inner {
        border: 0;
      }
      .ms-Fabric--isFocusVisible &:focus:after {
        border-radius: 0;
        border: 2px solid #1362ae;
        bottom: -3px;
        content: "";
        left: -3px;
        outline: transparent;
        position: absolute;
        right: -3px;
        top: -3px;
        z-index: 1;
      }
      @media screen and (-ms-high-contrast: active), (forced-colors: active){.ms-Fabric--isFocusVisible &:focus:after {
        bottom: -2px;
        left: -2px;
        outline-color: ButtonText;
        right: -2px;
        top: -2px;
      }
      &:active > * {
        left: 0px;
        position: relative;
        top: 0px;
      }
      @media screen and (-ms-high-contrast: active), (forced-colors: active){& {
        border-color: Window;
      }
      &:hover {
        color: #1362ae;
      }
      @media screen and (-ms-high-contrast: active), (forced-colors: active){&:hover {
        color: Highlight;
      }
      &:hover .ms-Button-icon {
        color: #1362ae;
      }
      &:active {
        color: rgba(11, 11, 11, 0.95);
      }
      &:active .ms-Button-icon {
        color: #0a365f;
      }
      &.ms-Button.ms-Button--action {
        color: #1362ae;
        height: auto;
        min-height: 40px;
        padding-bottom: 10px;
        padding-left: 4px;
        padding-right: 4px;
        padding-top: 10px;
        text-align: left;
      }
      & span {
        flex-direction: row;
      }
      &.ms-Button--action .ms-Button-icon {
        color: #1362ae;
        font-size: 1.25rem;
        transform: translateY(1px);
      }
      & .ms-Button-flexContainer {
        align-items: baseline;
      }
      &.ms-Button--action .ms-Button-label {
        line-height: 1.5;
      }
      &.ms-Button--action:hover .ms-Button-label {
        color: rgba(0, 0, 0, 0.95);
        text-decoration: underline;
      }
      &.ms-Button--action:hover .ms-Button-icon {
        color: rgba(0, 0, 0, 0.95);
      }
      &.ms-Button--action:active .ms-Button-label {
        text-align: left;
        text-decoration: none;
      }
      &.ms-Button--action:disabled {
        color: #bbbbbb;
      }
      &.ms-Button--action:disabled i {
        color: #bbbbbb;
      }
  data-is-focusable={true}
  onClick={[Function]}
  onKeyDown={[Function]}
  onKeyPress={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  type="button"
>
  <span
    className=
        ms-Button-flexContainer
        {
          align-items: center;
          display: flex;
          flex-wrap: nowrap;
          height: 100%;
          justify-content: flex-start;
        }
    data-automationid="splitbuttonprimary"
  />
</button>
`;
